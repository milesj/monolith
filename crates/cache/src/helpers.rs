use moon_logger::warn;
use std::env;
use std::time::SystemTime;

pub const LOG_TARGET: &str = "moon:cache";

static mut LOGGED_WARNING: bool = false;

pub fn get_cache_env_var() -> String {
    if let Ok(var) = env::var("MOON_CACHE") {
        if var == "off" || var == "read" || var == "write" {
            return var;
        }

        // We only want to show this once, not everytime the function is called
        unsafe {
            if !LOGGED_WARNING {
                LOGGED_WARNING = true;

                warn!(
                    target: LOG_TARGET,
                    "Unknown MOON_CACHE environment variable value \"{}\", falling back to write mode",
                    var
                );
            }
        }
    }

    String::from("write")
}

pub fn is_readable() -> bool {
    get_cache_env_var() != "off"
}

pub fn is_writable() -> bool {
    get_cache_env_var() == "write"
}

pub fn to_millis(time: SystemTime) -> u128 {
    match time.duration_since(SystemTime::UNIX_EPOCH) {
        Ok(d) => d.as_millis(),
        Err(_) => 0,
    }
}

#[cfg(test)]
pub async fn run_with_env<T, F, Fut>(env: &str, callback: F) -> T
where
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = T>,
{
    if env.is_empty() {
        env::remove_var("MOON_CACHE");
    } else {
        env::set_var("MOON_CACHE", env);
    }

    let result = callback().await;

    env::remove_var("MOON_CACHE");

    result
}
