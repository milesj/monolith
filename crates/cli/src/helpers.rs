use console::{set_colors_enabled, set_colors_enabled_stderr};
use indicatif::{ProgressBar, ProgressStyle};
use moon_terminal::create_theme;
use starbase_styles::color::{no_color, supports_color};
use std::env;
use std::time::Duration;

pub fn map_list<T, F>(list: &[T], fmt: F) -> String
where
    F: Fn(&T) -> String,
{
    list.iter().map(fmt).collect::<Vec<_>>().join(", ")
}

pub fn fully_qualify_version(version: &str) -> String {
    if version.is_empty() {
        return version.to_owned();
    }

    let mut parts = version.split('.');

    [
        parts.next().unwrap_or("0"),
        parts.next().unwrap_or("0"),
        parts.next().unwrap_or("0"),
    ]
    .join(".")
}

pub fn create_progress_bar<S: AsRef<str>, F: AsRef<str>>(start: S) -> impl FnOnce(F, bool) {
    let pb = ProgressBar::new_spinner();
    pb.set_message(start.as_ref().to_owned());
    pb.enable_steady_tick(Duration::from_millis(50));

    move |finish, passed| {
        let theme = create_theme();

        pb.set_style(
            ProgressStyle::default_spinner()
                .template("{prefix} {msg}")
                .unwrap(),
        );

        if passed {
            pb.set_prefix(theme.success_prefix.to_string());
        } else {
            pb.set_prefix(theme.error_prefix.to_string());
        }

        pb.finish_with_message(finish.as_ref().to_owned());
    }
}

fn setup_no_colors() {
    env::set_var("NO_COLOR", "1");
    // https://github.com/mitsuhiko/clicolors-control/issues/19
    env::set_var("CLICOLOR", "0");

    set_colors_enabled(false);
    set_colors_enabled_stderr(false);
}

pub fn setup_colors(force: bool) {
    // If being forced by --color or other env vars
    if force
        || env::var("MOON_COLOR").is_ok()
        || env::var("FORCE_COLOR").is_ok()
        || env::var("CLICOLOR_FORCE").is_ok()
    {
        let mut color_level = env::var("MOON_COLOR")
            .or_else(|_| env::var("FORCE_COLOR"))
            .unwrap_or("3".to_owned());

        // https://nodejs.org/api/cli.html#force_color1-2-3
        if color_level.is_empty() || color_level == "true" {
            color_level = "1".to_owned();
        } else if color_level == "false" {
            color_level = "0".to_owned();
        }

        if color_level == "0" {
            setup_no_colors();
        } else {
            set_colors_enabled(true);
            set_colors_enabled_stderr(true);

            // https://bixense.com/clicolors/
            env::set_var("CLICOLOR_FORCE", &color_level);
            env::set_var("FORCE_COLOR", &color_level);
        }

        return;
    }

    if no_color() {
        setup_no_colors();
    } else {
        env::set_var("CLICOLOR", supports_color().to_string());
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serial_test::serial;

    fn reset_vars() {
        env::remove_var("NO_COLOR");
        env::remove_var("CLICOLOR");
        env::remove_var("CLICOLOR_FORCE");
        env::remove_var("FORCE_COLOR");
        env::remove_var("MOON_COLOR");
    }

    mod setup_color {
        use super::*;

        mod no_color {
            use super::*;

            #[test]
            #[serial]
            fn sets_vars() {
                env::set_var("NO_COLOR", "1");

                setup_colors(false);

                assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                reset_vars();
            }
        }

        mod forced_color {
            use super::*;

            // #[test]
            // #[serial]
            // fn forces_via_arg() {
            //     setup_colors(true);

            //     assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "2");
            //     assert_eq!(env::var("FORCE_COLOR").unwrap(), "2");
            //     assert!(env::var("NO_COLOR").is_err());

            //     reset_vars();
            // }

            // #[test]
            // #[serial]
            // fn forces_over_no_color() {
            //     env::set_var("NO_COLOR", "1");

            //     setup_colors(true);

            //     assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "2");
            //     assert_eq!(env::var("FORCE_COLOR").unwrap(), "2");
            //     assert_eq!(env::var("NO_COLOR").unwrap(), "1");

            //     reset_vars();
            // }

            #[test]
            #[serial]
            fn disables_if_zero() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "0");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                    assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn disables_if_false_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "false");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                    assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn enables_if_empty_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "1");
                    assert_eq!(env::var("FORCE_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn enables_if_true_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "true");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "1");
                    assert_eq!(env::var("FORCE_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }
        }
    }
}
