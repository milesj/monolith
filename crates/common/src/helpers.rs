use std::sync::OnceLock;

pub fn enable_pkl_configs() {
    std::env::set_var("MOON_EXPERIMENT_PKL_CONFIG", "true");
}

pub fn supports_pkl_configs() -> bool {
    static PKL_CACHE: OnceLock<bool> = OnceLock::new();

    *PKL_CACHE.get_or_init(|| {
        std::env::var("MOON_EXPERIMENT_PKL_CONFIG")
            .is_ok_and(|value| value == "1" || value == "true")
    })
}
