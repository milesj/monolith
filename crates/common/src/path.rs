pub use relative_path::*;
use starbase_styles::color;
use std::path::Path;

// Named types for better readability
pub type ProjectRelativePath = RelativePath;
pub type ProjectRelativePathBuf = RelativePathBuf;
pub type WorkspaceRelativePath = RelativePath;
pub type WorkspaceRelativePathBuf = RelativePathBuf;

#[cfg(unix)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

#[cfg(windows)]
#[inline]
pub fn normalize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('/', "\\")
}

#[inline]
pub fn standardize_separators<T: AsRef<str>>(path: T) -> String {
    path.as_ref().replace('\\', "/")
}

pub enum RelativeFrom<'path> {
    Project(&'path str),
    Workspace,
}

pub fn expand_to_workspace_relative<P: AsRef<str>>(
    from_format: RelativeFrom,
    path: P,
) -> WorkspaceRelativePathBuf {
    let path = standardize_separators(path.as_ref());

    match from_format {
        RelativeFrom::Project(source) => {
            // Root-level project
            if source.is_empty() || source == "." {
                WorkspaceRelativePathBuf::from(path)

                // Project-level, prefix with source path
            } else {
                let project_source = standardize_separators(source);

                if let Some(negated_glob) = path.strip_prefix('!') {
                    WorkspaceRelativePathBuf::from(format!("!{project_source}")).join(negated_glob)
                } else {
                    WorkspaceRelativePathBuf::from(project_source).join(path)
                }
            }
        }
        RelativeFrom::Workspace => WorkspaceRelativePathBuf::from(path),
    }
}

#[inline]
pub fn to_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    let path = path.as_ref();

    match path.to_str() {
        Some(p) => Ok(p.to_owned()),
        None => Err(miette::miette!(
            "Path {} contains invalid UTF-8 characters.",
            color::path(path)
        )),
    }
}

#[inline]
pub fn to_virtual_string<T: AsRef<Path>>(path: T) -> miette::Result<String> {
    Ok(standardize_separators(to_string(path)?))
}

#[inline]
pub fn exe_name<T: AsRef<str>>(name: T) -> String {
    #[cfg(windows)]
    {
        format!("{}.exe", name.as_ref())
    }

    #[cfg(not(windows))]
    {
        name.as_ref().into()
    }
}
