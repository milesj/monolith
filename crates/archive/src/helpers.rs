use moon_error::{map_io_to_fs_error, MoonError};
use moon_utils::path;
use std::fs;
use std::path::{Path, PathBuf};

pub fn ensure_dir(dir: &Path) -> Result<(), MoonError> {
    if !dir.exists() {
        fs::create_dir_all(dir).map_err(|e| map_io_to_fs_error(e, dir.to_path_buf()))?;
    }

    Ok(())
}

pub fn prepend_name(name: &str, prefix: &str) -> String {
    if prefix.is_empty() {
        return name.to_owned();
    }

    // Use native path utils to join the paths, so we can ensure
    // the parts are joined correctly within the archive!
    let parts: PathBuf = [prefix, name].iter().collect();

    path::normalize(parts).to_string_lossy().to_string()
}
