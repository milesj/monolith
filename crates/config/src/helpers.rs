use crate::errors::ConfigError;
use moon_utils::{
    fs::{self, temp},
    path,
};
use std::collections::VecDeque;
use std::path::{Path, PathBuf};
use yaml_rust::{Yaml, YamlLoader};

pub fn download_and_cache_config(url: &str) -> Result<PathBuf, ConfigError> {
    let file = temp::get_file(url, "yml");

    if temp::read(&file)?.is_some() {
        return Ok(file);
    }

    let error_handler = |_| ConfigError::FailedDownload(url.to_owned());
    let data = reqwest::blocking::get(url)
        .map_err(error_handler)?
        .text()
        .map_err(error_handler)?;

    temp::write(&file, data)?;

    Ok(file)
}

pub fn gather_extended_sources<T: AsRef<Path>>(
    base_config: T,
) -> Result<Vec<PathBuf>, ConfigError> {
    let base_config = base_config.as_ref();
    let mut queue = VecDeque::from(vec![path::to_string(base_config)?]);
    let mut sources = vec![];

    while !queue.is_empty() {
        if let Some(config) = queue.pop_front() {
            let mut config_path = PathBuf::from(&config);

            if !config.ends_with(".yml") && !config.ends_with(".yaml") {
                return Err(ConfigError::UnsupportedExtendsDocument(config));
            }

            // For https, download the config and store it in the temp cache
            if config.starts_with("https") {
                config_path = download_and_cache_config(&config)?;
            } else if config.starts_with("http") {
                return Err(ConfigError::UnsupportedHttps(config));
            }

            // Otherwise we have a possible file path
            if !config_path.exists() {
                return Err(ConfigError::MissingFile(config));
            }

            sources.push(config_path.clone());

            // Parse the YAML document and attempt to extract the `extends` field.
            // We can't use serde here as the shape of the document may be invalid
            // or incomplete.
            let yaml = YamlLoader::load_from_str(&fs::sync::read(&config_path)?)
                .map_err(|e| ConfigError::InvalidYaml(config_path.clone(), e.to_string()))?;

            let doc = &yaml[0];

            // Field does not exist!
            if doc["extends"].is_badvalue() {
                continue;
            }

            match &doc["extends"] {
                Yaml::String(extends) => {
                    if extends.starts_with("http") {
                        queue.push_back(extends.to_owned());
                    } else {
                        queue.push_back(path::to_string(
                            config_path.parent().unwrap().join(extends),
                        )?);
                    }
                }
                _ => {
                    return Err(ConfigError::InvalidExtendsField);
                }
            }
        }
    }

    // Reverse the order as we must load the extended leaf first,
    // with the initial base config overriding as last.
    sources.reverse();

    Ok(sources)
}
