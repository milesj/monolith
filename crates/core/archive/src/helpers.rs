use moon_utils::path;
use std::path::PathBuf;

pub fn prepend_name(name: &str, prefix: &str) -> String {
    if prefix.is_empty() {
        return name.to_owned();
    }

    // Use native path utils to join the paths, so we can ensure
    // the parts are joined correctly within the archive!
    let parts: PathBuf = [prefix, name].iter().collect();

    path::normalize(parts).to_string_lossy().to_string()
}
