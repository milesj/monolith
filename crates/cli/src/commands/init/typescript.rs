use super::InitOptions;
use crate::helpers::AnyError;
use dialoguer::theme::ColorfulTheme;
use dialoguer::Confirm;
use moon_config::load_workspace_typescript_config_template;
use moon_node_lang::tsconfig::TsConfigJson;
use moon_terminal::label_header;
use std::path::Path;
use tera::{Context, Error, Tera};

fn render_template(context: Context) -> Result<String, Error> {
    Tera::one_off(load_workspace_typescript_config_template(), &context, false)
}

pub async fn init_typescript(
    dest_dir: &Path,
    options: &InitOptions,
    theme: &ColorfulTheme,
) -> Result<String, AnyError> {
    if !options.yes {
        println!("\n{}\n", label_header("TypeScript"));
    }

    let project_refs = if let Ok(Some(tsconfig)) = TsConfigJson::read(dest_dir) {
        match tsconfig.compiler_options {
            Some(co) => co.composite.unwrap_or_default(),
            None => tsconfig.references.is_some(),
        }
    } else {
        options.yes
            || Confirm::with_theme(theme)
                .with_prompt("Use project references?")
                .interact()?
    };

    let mut route_cache = false;
    let mut sync_paths = false;

    if project_refs {
        route_cache = options.yes
            || Confirm::with_theme(theme)
                .with_prompt("Route declaration output to moons cache?")
                .interact()?;

        sync_paths = options.yes
            || Confirm::with_theme(theme)
                .with_prompt("Sync project references as path aliases?")
                .interact()?;
    }

    let mut context = Context::new();
    context.insert("project_refs", &project_refs);
    context.insert("route_cache", &route_cache);
    context.insert("sync_paths", &sync_paths);

    Ok(render_template(context)?)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn renders_default() {
        let mut context = Context::new();
        context.insert("project_refs", &false);
        context.insert("route_cache", &false);
        context.insert("sync_paths", &false);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_project_refs() {
        let mut context = Context::new();
        context.insert("project_refs", &true);
        context.insert("route_cache", &false);
        context.insert("sync_paths", &false);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_route_cache() {
        let mut context = Context::new();
        context.insert("project_refs", &true);
        context.insert("route_cache", &true);
        context.insert("sync_paths", &false);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_sync_paths() {
        let mut context = Context::new();
        context.insert("project_refs", &true);
        context.insert("route_cache", &true);
        context.insert("sync_paths", &true);

        assert_snapshot!(render_template(context).unwrap());
    }
}
