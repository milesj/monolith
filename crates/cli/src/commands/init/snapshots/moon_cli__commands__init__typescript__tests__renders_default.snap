---
source: crates/cli/src/commands/init/typescript.rs
assertion_line: 69
expression: render_template(context).unwrap()
---
# Configures how moon integrates with TypeScript.
typescript:
  # When `syncProjectReferences` is enabled and a dependent project reference
  # *does not* have a `tsconfig.json`, automatically create one.
  createMissingConfig: false

  # Name of `tsconfig.json` file in each project root.
  # projectConfigFileName: 'tsconfig.json'

  # Name of `tsconfig.json` file in the workspace root.
  # rootConfigFileName: 'tsconfig.json'

  # Name of the config file in the workspace root that defines shared compiler
  # options for all project reference based config files.
  # rootOptionsConfigFileName: 'tsconfig.options.json'

  # Update a project's `tsconfig.json` to route the `outDir` compiler option
  # to moon's `.moon/cache` directory.
  routeOutDirToCache: false

  # Sync a project's `dependsOn` as project references within the
  # project's `tsconfig.json` and the workspace root `tsconfig.json`.
  syncProjectReferences: false

  # Sync a project's project references as import aliases to the `paths`
  # compiler option in each applicable project.
  syncProjectReferencesToPaths: false

